File: heads/python/src/main/python/hydra/dsl/meta/lib/pairs.py
"""Phantom-typed term DSL for the hydra.lib.pairs library."""

from hydra.phantoms import TTerm
from hydra.dsl.meta.phantoms import lib_primitive1 as primitive1, lib_primitive3 as primitive3


def bimap(f: TTerm, g: TTerm, p: TTerm) -> TTerm:
    """Map over both elements of a pair."""
    return primitive3(f, g, p)


def first(pair: TTerm) -> TTerm:
    """
    Get the first element of a pair.

    Type: (a, b) -> a
    """
    return primitive1(pair)


def second(pair: TTerm) -> TTerm:
    """
    Get the second element of a pair.

    Type: (a, b) -> b
    """
    return primitive1(pair)

File: heads/haskell/src/main/haskell/Hydra/Dsl/Meta/Lib/Pairs.hs
-- | Phantom-typed term DSL for the hydra.lib.pairs library

module Hydra.Dsl.Meta.Lib.Pairs where

import Hydra.Typed
import Hydra.Dsl.Meta.Phantoms
import qualified Hydra.Dsl.Terms as Terms
import Hydra.Sources.Libraries


-- | Map over both elements of a pair.
bimap :: TypedTerm (a -> c) -> TypedTerm (b -> d) -> TypedTerm (a, b) -> TypedTerm (c, d)
bimap = primitive3 _pairs_bimap

-- | Get the first element of a pair.
first :: TypedTerm (a, b) -> TypedTerm a
first = primitive1 _pairs_first

-- | Get the second element of a pair.
second :: TypedTerm (a, b) -> TypedTerm b
second = primitive1 _pairs_second
